### PR TITLE
IGNITE-16743 Implement thin client CDC streamer

### PR DESCRIPTION
**Pull request by NSAmelchev**

### CODE REVIEW

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/CdcEventsApplier.java [hunk 1]
```diff
@@ -23,165 +23,122 @@
...
-import static org.apache.ignite.internal.processors.cache.GridCacheUtils.EXPIRE_TIME_CALCULATE;
-import static org.apache.ignite.internal.processors.cache.GridCacheUtils.TTL_NOT_CHANGED;
+import static org.apache.ignite.internal.processors.cache.GridCacheUtils.UNDEFINED_CACHE_ID;
 
 /**
- * Contains logic to process {@link CdcEvent} and apply them to the provided by {@link #ignite()} cluster.
+ * Contains logic to process {@link CdcEvent} and apply them to the cluster.
  */
-public abstract class CdcEventsApplier {
+public abstract class CdcEventsApplier<K, V>  {
```
Comment (nizhikov) [inline modules/cdc-ext/src/main/java/org/apache/ignite/cdc/CdcEventsApplier.java:39]: `AbstractCdcEventsApplier`?

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/CdcEventsIgniteApplier.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+    /** Caches. */
+    private final IntMap<IgniteInternalCache<BinaryObject, BinaryObject>> ignCaches = new IntHashMap<>();
+
+    /**
+     * @param ignite Destination cluster.
+     * @param maxBatchSize Maximum batch size.
+     * @param log Logger.
+     */
+    public CdcEventsIgniteApplier(Ignite ignite, int maxBatchSize, IgniteLogger log) {
```
Comment (nizhikov) [inline modules/cdc-ext/src/main/java/org/apache/ignite/cdc/CdcEventsIgniteApplier.java:61]: Le'ts make constructor parameter type `IgniteEx`.

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/thin/CdcEventsIgniteClientApplier.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import org.apache.ignite.internal.util.typedef.T2;
+import org.apache.ignite.internal.util.typedef.internal.CU;
+
+/**
+ * Contains logic to process {@link CdcEvent} and apply them to the destination cluster by thin client.
+ *
+ * @see TcpClientCache#putAllConflict(Map)
+ * @see TcpClientCache#removeAllConflict(Map)
+ */
+public class CdcEventsIgniteClientApplier extends CdcEventsApplier<Object, T2<Object, GridCacheVersion>> {
```
Comment (nizhikov) [inline modules/cdc-ext/src/main/java/org/apache/ignite/cdc/thin/CdcEventsIgniteClientApplier.java:38]: `CdcEventsThinClientApplier`?

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/CdcEventsIgniteApplier.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    /** {@inheritDoc} */
+    @Override protected GridCacheDrInfo toValue(Object val, GridCacheVersion ver) {
+        CacheObject cacheObj;
+
+        if (val instanceof CacheObject)
+            cacheObj = (CacheObject)val;
+        else
+            cacheObj = new CacheObjectImpl(val, null);
+
+        return new GridCacheDrInfo(cacheObj, ver);
```
Comment (nizhikov) [inline modules/cdc-ext/src/main/java/org/apache/ignite/cdc/CdcEventsIgniteApplier.java:114]: Let's check here expiry policy for cache to eliminate object recreation on apply.

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/AbstractIgniteCdcStreamer.java [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    /**
+     * Sets cache names that participate in CDC.
+     *
+     * @param caches Cache names.
+     * @return {@code this} for chaining.
+     */
+    public T setCaches(Set<String> caches) {
+        this.caches = caches;
+
+        return (T)this;
```
Comment (nizhikov) [inline modules/cdc-ext/src/main/java/org/apache/ignite/cdc/AbstractIgniteCdcStreamer.java:172]: Let's return just `AbstractIgniteCdcStreamer` and eliminate generic type parameter in class definition.

File: modules/cdc-ext/src/test/java/org/apache/ignite/cdc/thin/CdcIgniteToIgniteClientReplicationTest.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+    /** {@inheritDoc} */
+    @Override protected void checkConsumerMetrics(Function<String, Long> longMetric) {
+        assertNotNull(longMetric.apply(IgniteToIgniteCdcStreamer.LAST_EVT_TIME));
+        assertNotNull(longMetric.apply(IgniteToIgniteCdcStreamer.EVTS_CNT));
+    }
+
+    /**
+     * @param srcCfg Ignite source node configuration.
+     * @param destCfg Ignite destination cluster configuration.
```
Comment (nizhikov) [inline modules/cdc-ext/src/test/java/org/apache/ignite/cdc/thin/CdcIgniteToIgniteClientReplicationTest.java:74]: There is no destCfg parameter.

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/CdcEventsApplier.java (hunk 1 above)
Comment (NSAmelchev) [reply modules/cdc-ext/src/main/java/org/apache/ignite/cdc/CdcEventsApplier.java]: Done.

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/CdcEventsIgniteApplier.java (hunk 2 above)
Comment (NSAmelchev) [reply modules/cdc-ext/src/main/java/org/apache/ignite/cdc/CdcEventsIgniteApplier.java]: Done.

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/thin/CdcEventsIgniteClientApplier.java (hunk 3 above)
Comment (NSAmelchev) [reply modules/cdc-ext/src/main/java/org/apache/ignite/cdc/thin/CdcEventsIgniteClientApplier.java]: I think `IgniteClient` is more applicable

File: modules/cdc-ext/src/test/java/org/apache/ignite/cdc/thin/CdcIgniteToIgniteClientReplicationTest.java (hunk 6 above)
Comment (NSAmelchev) [reply modules/cdc-ext/src/test/java/org/apache/ignite/cdc/thin/CdcIgniteToIgniteClientReplicationTest.java]: Done. Thanks.

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/AbstractIgniteCdcStreamer.java (hunk 5 above)
Comment (NSAmelchev) [reply modules/cdc-ext/src/main/java/org/apache/ignite/cdc/AbstractIgniteCdcStreamer.java:172]: We can't use the abstract streamer: implementation-specific configurations will not compile: 

```
new IgniteToIgniteCdcStreamer()
   .setMaxBatchSize(KEYS_CNT)
// Compile error. Expected `IgniteToIgniteCdcStreamer` that contains the requested method.
   .setDestinationIgniteConfiguration(destCfg)
   .setCaches(Collections.singleton(cache)));
```

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/AbstractIgniteCdcStreamer.java [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+     * @return {@code this} for chaining.
+     */
+    public AbstractIgniteCdcStreamer setMaxBatchSize(int maxBatchSize) {
+        this.maxBatchSize = maxBatchSize;
+
+        return this;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onMappings(Iterator<TypeMapping> mappings) {
```
Comment (nizhikov) [inline modules/cdc-ext/src/main/java/org/apache/ignite/cdc/AbstractIgniteCdcStreamer.java:192]: Please, move logic methods upper then get/set methods.

**Review by nizhikov — APPROVED** [review]

File: modules/cdc-ext/src/main/java/org/apache/ignite/cdc/AbstractIgniteCdcStreamer.java (hunk 7 above)
Comment (NSAmelchev) [reply modules/cdc-ext/src/main/java/org/apache/ignite/cdc/AbstractIgniteCdcStreamer.java:156]: Done.

**Pull request merged**